ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/StateView.hs

```diff
@@ -1,104 +1,219 @@
+
+instance (StandardHash blk, Show blk, Show (Header blk)) => Condense (PeerSimulatorResult blk) where
+  condense PeerSimulatorResult{psePeerId, pseResult} =
+    condense psePeerId ++ " " ++ condense pseResult
+
+instance (StandardHash blk, Show blk, Show (Header blk)) => CondenseList (PeerSimulatorResult blk) where
+  condenseList results =
+    zipWith
+      (\peerId result -> peerId ++ " " ++ result)
+      (padListWith PadRight $ map (show . psePeerId) results)
+      (padListWith PadRight $ map (condense . pseResult) results)
 
 -- | A state view is a partial view of the state of the whole peer simulator.
 -- This includes information about the part of the code that is being tested
 -- (for instance the fragment that is selected by the ChainDB) but also
 -- information about the mocked peers (for instance the exceptions raised in the
 -- mocked ChainSync server threads).
 data StateView blk = StateView {
-    svSelectedChain       :: AnchoredFragment (Header blk),
-    svChainSyncExceptions :: [ChainSyncException],
-    svTipBlock            :: Maybe blk
+    svSelectedChain        :: AnchoredFragment (Header blk),
+    svPeerSimulatorResults :: [PeerSimulatorResult blk],
+    svTipBlock             :: Maybe blk
   }
 
 instance Condense (StateView TestBlock) where
-  condense StateView {svSelectedChain, svChainSyncExceptions, svTipBlock} =
+  condense StateView {svSelectedChain, svPeerSimulatorResults, svTipBlock} =
     "SelectedChain: " ++ terseHFragment svSelectedChain ++ "\n"
     ++ "TipBlock: " ++ terseMaybe terseBlock svTipBlock ++ "\n"
-    ++ "ChainSyncExceptions:\n" ++ unlines (("  - " ++) . condense <$> svChainSyncExceptions)
+    ++ "PeerSimulatorResults:\n" ++ unlines (fmap ("  - " ++) $ condenseList $ sort svPeerSimulatorResults)
 
 -- | Return the list of peer ids for all peers whose ChainSync thread was killed
 -- by the node under test (that is it received 'ThreadKilled').
 chainSyncKilled :: StateView blk -> [PeerId]
 chainSyncKilled stateView =
   mapMaybe
-    (\ChainSyncException{csePeerId, cseException} ->
-       if wasKilled cseException
-         then Just csePeerId
+    (\PeerSimulatorResult{psePeerId, pseResult} ->
+       if wasKilled pseResult
+         then Just psePeerId
          else Nothing)
-    (svChainSyncExceptions stateView)
+    (svPeerSimulatorResults stateView)
   where
-    wasKilled :: SomeException -> Bool
-    wasKilled e = case fromException e of
-      Just ThreadKilled -> True
-      _                 -> False
+    wasKilled :: PeerSimulatorComponentResult blk -> Bool
+    wasKilled res =
+      let exnM = case res of
+            SomeChainSyncClientResult  (Left exn) -> Just exn
+            SomeChainSyncServerResult  (Left exn) -> Just exn
+            SomeBlockFetchClientResult (Left exn) -> Just exn
+            SomeBlockFetchServerResult (Left exn) -> Just exn
+            _                                     -> Nothing
```

Comment:
This is a tad repetitive. Can the exception be stored separately from the type of result?

Reply:
I'm not completely satisfied with the way `PeerSimulatorComponentResult blk` is structured either, but I haven't found any better alternative yet. There are two factors we must account for:
- When an exception is raised, we must register which component it comes from.
- The type of the (positive) result depends on the component it comes from.
I think what I've done is an okay-ish compromise, but I agree it leads to some repetition...